Sources/Helpers/OAWikiArticleHelper.mm

```diff
@@ -116,5 +120,61 @@ + (void) warnAboutExternalLoad:(NSString *)url
     [[OARootViewController instance] presentViewController:alert animated:YES completion:nil];
 }
 
++ (NSString *) getFirstParagraph:(NSString *)descriptionHtml
+{
+   if (descriptionHtml)
+   {
+       NSString *firstParagraph = [self.class getPartialContent:descriptionHtml];
+       if (firstParagraph && firstParagraph.length > 0)
+           return firstParagraph;
+   }
+   return descriptionHtml;
+}
+
++ (NSString *) getPartialContent:(NSString *)source
+{
+    if (!source || source.length == 0)
+        return nil;
+    
+    NSString *content = [source regexReplacePattern:@"\n" newString:@""];
+    int firstParagraphStart = [content indexOf:kPOpened];
+    int firstParagraphEnd = [content indexOf:kPClosed];
+    firstParagraphEnd = firstParagraphEnd < firstParagraphStart ? [content indexOf:kPClosed start:firstParagraphStart] : firstParagraphEnd;
+    NSString *firstParagraphHtml = nil;
+    if (firstParagraphStart != -1 && firstParagraphEnd != -1 && firstParagraphEnd >= firstParagraphStart)
+    {
+        firstParagraphHtml = [content substringWithRange:NSMakeRange(firstParagraphStart, firstParagraphEnd - firstParagraphStart + kPClosed.length)];
+        while (([[firstParagraphHtml substringWithRange:NSMakeRange(kPOpened.length, firstParagraphHtml.length - kPOpened.length - kPClosed.length)] trim].length == 0
+               && (firstParagraphEnd + kPClosed.length < content.length))
+               || [[firstParagraphHtml regexReplacePattern:@"(<a.+?/a>)|(<div.+?/div>)" newString:@""] trim].length == 0)
+        {
+            firstParagraphStart = [content indexOf:kPOpened start:firstParagraphEnd];
+            firstParagraphEnd = firstParagraphStart == -1 ? -1 : [content indexOf:kPClosed start:firstParagraphStart];
+            if (firstParagraphStart != -1 && firstParagraphEnd != -1)
+                firstParagraphHtml = [content substringWithRange:NSMakeRange(firstParagraphStart, firstParagraphEnd - firstParagraphStart + kPClosed.length)];
+            else
+                break;
+        }
+    }
+    
+    if (!firstParagraphHtml || firstParagraphHtml.length == 0)
+        firstParagraphHtml = source;
+    if (!firstParagraphHtml || firstParagraphHtml.length == 0)
+        return nil;
+    
+    NSString *firstParagraphText = [[firstParagraphHtml regexReplacePattern:@"(<(/)(a|img)>)|(<(a|img).+?>)|(<div.+?/div>)" newString:@""] stringByReplacingOccurrencesOfString:@"<br>" withString:@""];
```

Comment:
Are single slashes like / correct in obj-c?

Reply:
regex results look the same
[screenshot](https://www.dropbox.com/s/isdehf5c884f6e9/Screenshot%202022-01-27%20at%2014.21.39.png?dl=0)